p2panda-rs/src/storage_provider/traits/entry_store.rs

```diff
@@ -15,6 +15,88 @@ use crate::storage_provider::traits::AsStorageEntry;
+        &self,
+        entry: &StorageEntry,
+    ) -> Result<Option<StorageEntry>, EntryStorageError> {
+        if entry.seq_num().is_first() {
+            return Ok(None);
+        };
+
+        // Unwrap as we know this isn't the first sequence number because of the above condition
+        let backlink_seq_num = SeqNum::new(entry.seq_num().as_u64() - 1).unwrap();
+        let expected_backlink = self
+            .entry_at_seq_num(&entry.author(), &entry.log_id(), &backlink_seq_num)
+            .await?
+            .ok_or_else(|| EntryStorageError::ExpectedBacklinkMissing(entry.hash()))?;
+
+        // compare the expected backlink hash and the stated backlink hash
+        if expected_backlink.hash() != entry.backlink_hash().unwrap() {
+            return Err(EntryStorageError::InvalidBacklinkPassed(entry.hash()));
+        }
+        Ok(Some(expected_backlink))
+    }
+
+    /// Retrieve and verify the skiplink of a passed entry.
+    ///
+    /// Fetches the expected skiplink for the passed entry (based on SeqNum, Author & LogId), if
+    /// it is found it verifies its hash against the skiplink entry hash encoded in the passed
+    /// entry.
+    ///
+    /// If either the expected skiplink is not found in storage, or it doesn't match the oneencoded in
+    /// the passed entry, then an error is returned. If no skiplink is required for an entry at this
+    /// seq num, and it wasn't encoded with one, then `None` is returned.
+    ///  
+    /// If the skiplink is retrieved and validated against the encoded entries skiplink successfully
+    /// the skiplink entry is returned.
+    async fn try_get_skiplink(
+        &self,
+        entry: &StorageEntry,
+    ) -> Result<Option<StorageEntry>, EntryStorageError> {
+        // If a skiplink isn't required and it wasn't provided, return already now
+        if !is_lipmaa_required(entry.seq_num().as_u64()) && entry.skiplink_hash().is_none() {
+            return Ok(None);
+        };
+
+        // Derive the expected skiplink seq number from this entries sequence number
+        let expected_skiplink = match entry.seq_num().skiplink_seq_num() {
+            // Retrieve the expected skiplink from the database
+            Some(seq_num) => {
+                let expected_skiplink_entry = self
+                    .entry_at_seq_num(&entry.author(), &entry.log_id(), &seq_num)
+                    .await?
+                    .ok_or_else(|| EntryStorageError::ExpectedSkiplinkMissing(entry.hash()))?;
+                Some(expected_skiplink_entry)
+            }
+            // Or if there is no skiplink for entries at this sequence number return None
+            None => None,
+        };
+
+        // compare the expected skiplink hash and the stated skiplink hash
+        if expected_skiplink.clone().map(|entry| entry.hash()) != entry.skiplink_hash() {
+            return Err(EntryStorageError::InvalidSkiplinkPassed(entry.hash()));
+        }
+        Ok(expected_skiplink)
+    }
```

Comment:
I'd like a little feedback on this method. I see two possible approaches (maybe there are more?).

1) Fetch the expected skiplink for an entry from storage based on `Author`, `LogId` and `SeqNum`, if it is found in storage we then validate it's hash against the one encoded in the db.

2) Fetch the entry for an entries' encoded skiplink by it's hash, compare it's `Author` `SeqNum` and `LogId` to the expected values.

Option 1) is what we are currently doing, option 2) is a different approach. The reason I wonder about option 2) is because I believe in most implementations queries for fetching an entry by it's hash will be more efficient (this is an indexed value in the db). Rather than scanning across author, log_id and seq_num as would be the case I think in option 1). When we implement `get_all_lipmaa_entries_for_entry()` performance may be more of an issue.

Any thoughts on this anyone? Maybe @pietgeursen has particular insight ;-p ?

Reply:
> Rather than scanning across author, log_id and seq_num

At least in aquadoggo, you have: `PRIMARY KEY (author, log_id, seq_num)` which should make it very fast to search for them.

> When we implement get_all_lipmaa_entries_for_entry() performance may be more of an issue.

Are you going to try and implement `get_all_lipmaa_entries_for_entry` by calling `try_get_skiplink` multiple times?

I'd rather not do that for performance reasons. I think that can be done in a single query like:

(probably incorrect sql)
```sql
SELECT * FROM entries WHERE author = <> AND log_id = <> AND seq_num IN (1,2,3 ...)
```

Reply:
Thanks, yeh, you're right, in aquadoggo we are covered.

re: `get_all_lipmaa_entries_for_entry` I was planning to do something like your psuedo code 👍

You don't see any downside to option 2) then? I'm mostly just curious 😄

Reply:
I just made [this PR](https://github.com/pietgeursen/lipmaa-link/pull/3) in `lipmaa_link`. Do you reckon that's be handy for this?

Reply:
> You don't see any downside to option 2) then?

I've had one beer so I'm not thinking super clearly. But does one give us an advantage in detecting forks? Maybe not? Idk lol.

Reply:
Haha, I've had one too many Onsen (and beers) to be thinking clearly 🤣

But that's the kind of thing I was wondering about. I can't quite tell if the two methods are exactly the same, or subtly different....

Reply:
> Do you reckon that's be handy for get_all_lipmaa_entries_for_entry?

Ahh, thanks, yes, useful.